sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,1045 @@
+         * the 'isRetryInitiated' flag ensures those drain loop iterations (those also see the mediator as terminated
+         * and drained) will not initiate duplicate retries.
+         */
+        volatile boolean isRetryInitiated;
+        /**
+         * The queue holding messages from the backing receiver's message publisher, waiting to be drained by
+         * the drain-loop iterations.
+         */
+        final Queue<Message> queue;
+
+        /**
+         * Create a mediator to channel events (messages, termination) from a receiver to recoverable-receiver.
+         *
+         * @param parent the recoverable-receiver (a.k.a. parent).
+         * @param receiver the receiver backing the mediator.
+         * @param prefetch the number of messages to prefetch using the receiver (for a less chatty network
+         *                 and faster message processing on the client).
+         * @param creditFlowMode the mode indicating how to compute the credit and when to send it to the broker.
+         */
+        ReactorReceiverMediator(RecoverableReactorReceiver parent, AmqpReceiveLink receiver, int prefetch,
+            CreditFlowMode creditFlowMode, ClientLogger logger) {
+            this.parent = parent;
+            this.receiver = receiver;
+            this.prefetch = prefetch;
+            this.creditFlowMode = creditFlowMode;
+            this.logger = logger;
+            // Obtain a resizable single-producer & single-consumer queue (SpscLinkedArrayQueue).
+            this.queue = Queues.<Message>get(Integer.MAX_VALUE).get();
+        }
+
+        /**
+         * Invoked by the parent {@link RecoverableReactorReceiver} when it is ready to use this new mediator
+         * (The mediator facilitate communication between the parent and the new receiver ({@link AmqpReceiveLink}) that
+         * mediator wraps). In response, this mediator notifies the parent about its readiness by invoking
+         * {@link RecoverableReactorReceiver#onMediatorReady(BiFunction)}.
+         */
+        void onParentReady() {
+            updateLogWithReceiverId(logger.atWarning()).log("Setting next mediator and waiting for activation.");
+            // 1. Subscribe for the messages on the Receiver (AmqpReceiveLink).
+            receiver.receive().subscribe(this);
+            // 2. Subscribe for the readiness (active) and terminal event on the Receiver (AmqpReceiveLink) endpoint.
+            final Disposable endpointDisposable = receiver.getEndpointStates()
+                .filter(s -> s == AmqpEndpointState.ACTIVE)
+                // ^ Pass down only readiness (active) event and terminal events.
+                .publishOn(ReceiversPumpingScheduler.instance())
+                // ^ Offload any initial drain (upon readiness) and final drain (upon termination) to ReceiversPumpingScheduler.
+                .doOnEach(event -> {
```

Comment:
IIRC we discussed if we should map this to an actual operator like .handle rather than a side-effect operator like doOn\*

Reply:
Reviewing the code again, it looks like we can simply remove `doOnEach(..)` and have `subscribe(,,,)` takes care of signal reactions. I’ll update the code.